skiko/src/awtMain/cpp/windows/directXSwingRedrawer.cc

```diff
@@ -0,0 +1,400 @@
+        if (!SUCCEEDED(CreateDXGIFactory1(IID_PPV_ARGS(&deviceFactory)))) {
+            return 0;
+        }
+        if (adapterPtr == 0) {
+            return 0;
+        }
+        gr_cp<IDXGIAdapter1> adapter((IDXGIAdapter1 *) adapterPtr);
+
+        D3D_FEATURE_LEVEL maxSupportedFeatureLevel = D3D_FEATURE_LEVEL_12_0;
+        D3D_FEATURE_LEVEL featureLevels[] = {
+            // TODO add D3D_FEATURE_LEVEL_12_2
+            D3D_FEATURE_LEVEL_12_1,
+            D3D_FEATURE_LEVEL_12_0
+        };
+
+        for (int i = 0; i < _countof(featureLevels); i++) {
+            if (SUCCEEDED(D3D12CreateDevice(adapter.get(), featureLevels[i], _uuidof(ID3D12Device), nullptr))) {
+                maxSupportedFeatureLevel = featureLevels[i];
+                break;
+            }
+        }
+
+        gr_cp<ID3D12Device> device;
+        if (!SUCCEEDED(D3D12CreateDevice(adapter.get(), maxSupportedFeatureLevel, IID_PPV_ARGS(&device)))) {
+            return 0;
+        }
+
+        // Create the command queue
+        gr_cp<ID3D12CommandQueue> queue;
+        D3D12_COMMAND_QUEUE_DESC queueDesc = {};
+        queueDesc.Flags = D3D12_COMMAND_QUEUE_FLAG_NONE;
+        queueDesc.Type = D3D12_COMMAND_LIST_TYPE_DIRECT;
+
+        if (!SUCCEEDED(device->CreateCommandQueue(&queueDesc, IID_PPV_ARGS(&queue)))) {
+            return 0;
+        }
+
+        ID3D12CommandAllocator* commandAllocator;
+        if (!SUCCEEDED(device->CreateCommandAllocator(D3D12_COMMAND_LIST_TYPE_DIRECT, IID_PPV_ARGS(&commandAllocator)))) {
+            return 0;
+        }
+
+        ID3D12GraphicsCommandList* commandList;
+        if (!SUCCEEDED(device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, commandAllocator, nullptr, IID_PPV_ARGS(&commandList)))) {
+            return 0;
+        }
+
+        ID3D12Fence* fence;
+        if (!SUCCEEDED(device->CreateFence(0, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)))) {
+            return 0;
+        }
+
+        HANDLE fenceEvent = CreateEventEx(nullptr, false, false, EVENT_ALL_ACCESS);
+        if (!fenceEvent) {
+            return 0;
+        }
+
+        DirectXOffscreenDevice *d3dDevice = new DirectXOffscreenDevice();
+        d3dDevice->texture = nullptr;
+        d3dDevice->readbackBuffer = nullptr;
+        d3dDevice->commandAllocator = commandAllocator;
+        d3dDevice->commandList = commandList;
+        d3dDevice->fence = fence;
+        d3dDevice->fenceEvent = fenceEvent;
+        d3dDevice->backendContext.fAdapter = adapter;
+        d3dDevice->backendContext.fDevice = device;
+        d3dDevice->backendContext.fQueue = queue;
+        d3dDevice->backendContext.fProtectedContext = GrProtected::kNo;
+
+        auto& textureDesc = d3dDevice->textureDesc;
+        textureDesc.Dimension = D3D12_RESOURCE_DIMENSION_TEXTURE2D;
+        textureDesc.Alignment = 0;
+        textureDesc.Width = 0;
+        textureDesc.Height = 0;
+        textureDesc.DepthOrArraySize = 1;
+        textureDesc.MipLevels = 1;
```

Comment:
I don't think that we need it

Reply:
I'd better keep it in explicit initial state to display that they are set somewhere else.

Reply:
It is needed 

https://learn.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_resource_desc